opentelemetry-sdk/src/trace/provider.rs

```diff
@@ -114,11 +147,41 @@ impl TracerProvider {
             .map(|processor| processor.force_flush())
             .collect()
     }
+
+    /// Shuts down the current `TracerProvider`.
+    ///
+    /// Note that shut down doesn't means the TracerProvider has dropped
+    pub fn shutdown(&self) -> TraceResult<()> {
+        if self
+            .is_shutdown
+            .compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst)
+            .is_ok()
+        {
+            // propagate the shutdown signal to processors
+            // it's up to the processor to properly block new spans after shutdown
+            let mut errs = vec![];
+            for processor in &self.inner.processors {
+                if let Err(err) = processor.shutdown() {
+                    errs.push(err);
+                }
+            }
+
+            if errs.is_empty() {
+                Ok(())
+            } else {
+                Err(TraceError::Other(format!("{errs:?}").into()))
+            }
+        } else {
+            Err(TraceError::Other(
+                "tracer provider already shut down".into(),
```

Comment:
We should probably use `Mutex` or `RwLock` instead of atomic variable to track shutdown. With the current setup, if there are two threads that are racing to shutdown the tracerprovider, the thread which fails the compare and swap and immediately come to the else condition. It wouldn't return an error message saying "tracer provider already shut down" without even waiting or verifying that the other thread has indeed completed the shutdown.

Reply:
Will rephrase the error to say "tracer provider is already shutting down or has been shut down" be useful here, instead of using the Mutex/RwLock ?

Reply:
It depends on what kind of experience we are targeting. I feel that shutting down a tracer provider does not have to be a perf optimized operation so it's okay to use locks. Shutdown would anyway not be a frequent scenario. I also like the clear status that locking offers about whether the provider is shutdown or not.

Reply:
The mutex/rwlock/atomic - is checked in hot path, so it needs to be performant!

Reply:
https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-sdk/src/logs/log_processor.rs#L93-L97 Logs. 
Looks like we are inconsistent here. so that need to be taken care as well!

Reply:
> The mutex/rwlock/atomic - is checked in hot path, so it needs to be performant!

Got it! In that case, we should make use of `Arc<Cell<bool>>` to track `is_shutdown` instead of atomics as that should offer better performance.

Reply:
why `Arc<Cell<bool>>` - this can cause race condition and data corruption right, as it's not thread safe.

Reply:
My bad! I thought we could use `Mutex` in conjunction with `Arc<Cell<T>>` to save the atomic read operation on hot-path. Rust wouldn't allow that.

Reply:
`Cell<T>` is not `Send` and `Sync` so compiler won't allow this